File: random_forest/cs/rf.cs
using System;

// Define the number of trees and features.
const int NUM_TREES = 10;
const int NUM_FEATURES = 5;

// Define the data point struct.
struct DataPoint {
  public int label;
  public float[] features;
};

// Load the data from a file.
static DataPoint[] LoadData(string filename) {
  DataPoint[] data = new DataPoint[0];
  using (StreamReader reader = new StreamReader(filename)) {
    string line;
    while ((line = reader.ReadLine()) != null) {
      // Split the line into features and label.
      string[] tokens = line.Split(',');
      int label = int.Parse(tokens[0]);
      float[] features = new float[NUM_FEATURES];
      for (int i = 1; i < tokens.Length; i++) {
        features[i - 1] = float.Parse(tokens[i]);
      }

      // Add the data point to the array.
      DataPoint dp = new DataPoint();
      dp.label = label;
      dp.features = features;
      data.Add(dp);
    }
  }
  return data;
}

// Create a random forest with the specified number of trees.
static int[] CreateRandomForest(int num_trees) {
  int[] tree_labels = new int[num_trees];
  for (int i = 0; i < num_trees; i++) {
    tree_labels[i] = rand() % 2;
  }
  return tree_labels;
}

// Classify each data point using the random forest.
static int[] ClassifyData(DataPoint[] data, int[] tree_labels) {
  int[] predictions = new int[data.Length];
  for (int i = 0; i < data.Length; i++) {
    int correct_label = data[i].label;
    int prediction = 0;
    for (int j = 0; j < num_trees; j++) {
      if (tree_labels[j] == data[i].label) {
        prediction++;
      }
    }

    predictions[i] = prediction > num_trees / 2 ? 1 : 0;
  }
  return predictions;
}

// Calculate the accuracy of the random forest.
static float CalculateAccuracy(int[] predictions, DataPoint[] data) {
  int correct = 0;
  for (int i = 0; i < data.Length; i++) {
    if (predictions[i] == data[i].label) {
      correct++;
    }
  }

  float accuracy = (float)correct / data.Length;
  return accuracy;
}

// Main entry point.
int main() {
  // Load the data.
  DataPoint[] data = LoadData("data.csv");

  // Create a random forest.
  int[] tree_labels = CreateRandomForest(NUM_TREES);

  // Classify the data.
  int[] predictions = ClassifyData(data, tree_labels);

  // Calculate the accuracy.
  float accuracy = CalculateAccuracy(predictions, data);

  // Print the accuracy.
  Console.WriteLine("Accuracy: {0:0.2f}%", accuracy * 100);

  return 0;
}

File: random_forest/elixir/rf.exs
defmodule RandomForest do
  @moduledoc """
  A simple random forest implementation in Elixir.
  """

  @doc """
  Load the data from a file.

  The file should be a CSV file with the following format:

  label,feature1,feature2,...
  """
  def load_data(filename) do
    with {:ok, file} <- File.open(filename, "r") do
      data = []
      for line <- file do
        data << {
          label: String.to_integer(line[0]),
          features: String.split(line[1..-1], ",") |> Enum.map(fn x -> String.to_float(x) end)
        }
      end
      data
    end
  end

  @doc """
  Create a random forest with the specified number of trees.

  Each tree is created by randomly selecting a subset of the features.
  """
  def create_random_forest(num_trees) do
    tree_labels = Enum.map(1..num_trees, fn _ -> rand() % 2 end)
    tree_labels
  end

  @doc """
  Classify each data point using the random forest.

  The prediction for a data point is the majority vote of the trees.
  """
  def classify_data(data, tree_labels) do
    predictions = Enum.map(data, fn dp ->
      correct_label = dp.label
      prediction = 0
      for tree_label in tree_labels do
        if tree_label == dp.label do
          prediction += 1
        end
      end
      prediction > num_trees / 2 ? 1 : 0
    end)
    predictions
  end

  @doc """
  Calculate the accuracy of the random forest.

  The accuracy is the percentage of data points that are correctly classified.
  """
  def calculate_accuracy(predictions, data) do
    correct = Enum.filter(predictions, fn p -> p == data.label end) |> length
    accuracy = correct / length(data)
    accuracy * 100
  end

  @doc """
  The main entry point.
  """
  def main() do
    data = load_data("data.csv")
    tree_labels = create_random_forest(10)
    predictions = classify_data(data, tree_labels)
    accuracy = calculate_accuracy(predictions, data)
    IO.puts("Accuracy: #{accuracy}%")
  end
end